ADD: [gles] BOB deinterlacing

koying opened the pull request.

koying committed:
FIX: [stf] de-invert y at source

koying committed:
ADD: [gles] BOB deinterlacing

FernetMenta commented:
awesome!

koying commented:
@FernetMenta Thanks for your tips :)
Test builds coming

fritsch commented:
awesome! @smallint @wolfgar ping!

smallint commented:
I keep that in mind. If someone wants to test it with current mainline, I am interested in results. As we are focusing currently on getting rid of GL for rendering, we don't use shaders anymore. One could enable switching between both methods later, though. Thanks for this.

koying committed:
CHG: [gles] use shader for generic yuv bob deinterlacing

GLES 2 does not support GL_UNPACK_ROW_LENGTH. That makes de-deinterlacing
a texture in CPU unbearably slow on some SoC's.

koying committed:
FIX: [gles] enable generic bob deinterlacing

koying committed:
ADD: [gles] enable contrast & brightness controls

koying commented:
I've added BOB for the yub2rgb path.

The easy way is 72b68761d3c178c69db8345a12fa17fd96b48ed3 , but the fact that GL_UNPACK_ROW_LENGTH is not implemented in GLES 2 has a big impact, so I added c851660faad618ef2aee0dee1f7e36893501722b to reuse an existing shader which apparently fell in disuse..

GL_UNPACK_ROW_LENGTH is implemented in GLES 3, though, but it's not quite mainstream yet. The alternative would thus be to go with the "standard" version and switch to GLES 3 in due time.

I've also reworked the shaders to remove the "if", which is apparently is less efficient in GLSL than doing unnecessary blending.

While playing with shaders, I also implemented contrast/brightness for gles. Is it ok to add to this PR? Otherwise, it will wait for this one to go in to avoid rebase hell.

fritsch commented:
Add them as separate commits to this PR I'd say.

Do you have an overview of which platforms have proper GLES3 support yet? Can those two be "#fidefed" somehow - so that we can use the best out of both worlds?

FernetMenta commented:
Im am fine with added commits to this PR too. Any issues in this area will be forwarded to the desk of our new shader expert anyway :)

koying commented:
Re GLES3, see, e.g. http://en.wikipedia.org/wiki/Mali_(GPU)
Mali 4XX is the vast majority of GPU around atm, and does not support it.

But yeah, I could implement GLES3 and fallback to GLES2.

fritsch commented:
That would be awesome. Performance wise - how fast works your implementation? Do we get 50 fps out of it?

koying commented:
Performances obviously depend on the GPU, but I get 50fps out of an AFTV Stick, which is quite low end, both on HD (accelerated) and SD (ffmpeg).

fritsch commented:
That's very nice! If you need some testing samples: http://solidrun.maltegrosse.de/~fritsch/

This PR is ready when you are :-)

Jalle19 commented on xbmc/cores/VideoRenderers/LinuxRendererGLES.cpp:
```diff
@@ -786,21 +803,32 @@ void CLinuxRendererGLES::LoadShaders(int field)
+            m_pYUVBobShader->Free();
+            delete m_pYUVBobShader;
+            m_pYUVBobShader = NULL;
+          }
```
This code is now duplicated in three places, might as well make it a method

koying replied:
The whole file deserves a big refactoring, really...
Plenty of duplication between the various render methods.

koying commented:
jenkins build this please

@fritsch If you want to test deinterlacing, this is the tricky sample to use ;)
http://www.semperpax.be/owncloud/public.php?service=files&t=6cc4e6fc01412e346a9ebf173a2ceca4

fritsch commented:
Hehe - thanks much. I always looked at the burosch samples, which are also
quite nice. But yours rocks more :-)

2015-01-20 14:20 GMT+01:00 Chris Browet <notifications@github.com>:

> jenkins build this please
>
> @fritsch <https://github.com/fritsch> If you want to test deinterlacing,
> this is the tricky sample to use ;)
>
> http://www.semperpax.be/owncloud/public.php?service=files&t=6cc4e6fc01412e346a9ebf173a2ceca4
>
> --
> Reply to this email directly or view it on GitHub
> <https://github.com/xbmc/xbmc/pull/6169#issuecomment-70651683>.
>



-- 
                   Key-ID:     0x1A995A9B
                   keyserver: pgp.mit.edu
==============================================================
Fingerprint: 4606 DA19 EC2E 9A0B 0157  C81B DA07 CF63 1A99 5A9B

koying commented:
Unrelated win32 failure, I guess? (http://jenkins.kodi.tv/job/WIN-32/3244/console)
@Memphiz ?

Will merge in 24h if no further comments.

Memphiz commented:
i think that slave has network issues ...

koying commented:
jenkins build and merge
